File: extract_auth.py
import os
import re
import glob

def parse_controllers():
    controller_files = []
    for root, dirs, files in os.walk('.'):
        for file in files:
            if file.endswith('Controller.cs'):
                controller_files.append(os.path.join(root, file))
                
    markdown_content = "# Backend API Authorization Analysis\n\n"
    markdown_content += "This document lists all API endpoints across the microservices, detailing their required permissions and roles.\n\n"

    for file_path in controller_files:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        class_match = re.search(r'class\s+(\w+Controller)', content)
        if not class_match:
            continue
        class_name = class_match.group(1)

        # Get service name from path (e.g., CynapCRM.Services.DocAPI)
        service_name = "Unknown Service"
        path_parts = file_path.split(os.sep)
        for part in path_parts:
            if "API" in part and part.startswith("CynapCRM"):
                service_name = part
                break

        class_route_match = re.search(r'\[Route\("([^"]+)"\)\][\s\S]*?class\s+' + class_name, content)
        class_route = class_route_match.group(1) if class_route_match else ""

        class_auth_match = re.search(r'\[Authorize(.*?)\][\s\S]*?class\s+' + class_name, content)
        class_auth = "[Authorize" + class_auth_match.group(1) + "]" if class_auth_match else ""
        
        class_anon_match = re.search(r'\[AllowAnonymous\][\s\S]*?class\s+' + class_name, content)
        if class_anon_match:
            class_auth = "[AllowAnonymous]"

        markdown_content += f"## {class_name} ({service_name})\n"
        if class_route:
            markdown_content += f"**Base Route:** `{class_route}`\n\n"
        markdown_content += f"**Class Level Authorization:** `{class_auth if class_auth else 'None'}`\n\n"
        
        markdown_content += "| Method | HTTP Verb | Route | Authorization | Access Control Explanation |\n"
        markdown_content += "|---|---|---|---|---|\n"

        lines = content.split('\n')
        current_http_verb = ""
        current_route = ""
        current_auth = ""
        in_method_attrs = False
        
        for i, line in enumerate(lines):
            line = line.strip()
            if line.startswith("[Http"):
                in_method_attrs = True
                verb_match = re.search(r'\[(HttpGet|HttpPost|HttpPut|HttpDelete|HttpPatch)', line)
                if verb_match:
                    current_http_verb = verb_match.group(1).replace("Http", "").upper()
                
                route_match = re.search(r'\("([^"]+)"\)', line)
                if route_match:
                    current_route = route_match.group(1)
            elif in_method_attrs and line.startswith("[Authorize"):
                current_auth = line
            elif in_method_attrs and line.startswith("[AllowAnonymous]"):
                current_auth = "[AllowAnonymous]"
            elif line.startswith("public") and "class" not in line and "interface" not in line and "(" in line:
                if in_method_attrs:
                    name_match = re.search(r'public\s+(?:async\s+)?(?:Task(?:<.*?>)?|IActionResult|ActionResult(?:<.*?>)?|\w+)\s+(\w+)\(', line)
                    if name_match:
                        method_name = name_match.group(1)
                        
                        final_auth = current_auth if current_auth else (class_auth if class_auth else "None")
                        
                        full_route = class_route
                        if current_route:
                            if not full_route.endswith('/'):
                                full_route += '/'
                            full_route += current_route
                            
                        explanation = ""
                        if final_auth == "[AllowAnonymous]":
                            explanation = "Public access, no token required."
                        elif final_auth == "[Authorize]":
                            explanation = "Requires valid authentication token."
                        elif "Roles" in final_auth:
                            roles_match = re.search(r'Roles\s*=\s*"([^"]+)"', final_auth)
                            if roles_match:
                                explanation = f"Requires valid token AND user must have one of these roles: **{roles_match.group(1)}**."
                        elif final_auth == "None":
                            explanation = "No explicit authorization rule found (might inherit global or fail open)."
                            
                        markdown_content += f"| {method_name} | {current_http_verb} | `{full_route}` | `{final_auth}` | {explanation} |\n"
                
                # Reset for next method
                current_http_verb = ""
                current_route = ""
                current_auth = ""
                in_method_attrs = False

        markdown_content += "\n---\n\n"

    with open("backend_auth_analysis.md", "w", encoding="utf-8") as f:
        f.write(markdown_content)

if __name__ == '__main__':
    parse_controllers()

File: ExtractAuth.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        var files = Directory.GetFiles(".", "*Controller.cs", SearchOption.AllDirectories);
        using var writer = new StreamWriter("AUTHORIZATION_ANALYSIS.md");
        writer.WriteLine("# API Authorization Analysis");
        writer.WriteLine();

        foreach (var file in files)
        {
            var content = File.ReadAllText(file);
            var classNameMatch = Regex.Match(content, @"class\s+(\w+Controller)");
            if (!classNameMatch.Success) continue;
            
            var className = classNameMatch.Groups[1].Value;
            var classRouteMatch = Regex.Match(content, @"\[Route\(""(.*?)""\)\]\s*(?:\[[^\]]+\]\s*)*public\s+class\s+" + className);
            var classRoute = classRouteMatch.Success ? classRouteMatch.Groups[1].Value : "";

            var classAuthMatch = Regex.Match(content, @"\[Authorize(?:.*?\)\]|\])\s*(?:\[[^\]]+\]\s*)*public\s+class\s+" + className);
            var classAuth = "";
            if (classAuthMatch.Success)
            {
                var matchVal = Regex.Match(classAuthMatch.Value, @"\[Authorize(.*?)\]");
                classAuth = matchVal.Success ? matchVal.Value : "[Authorize]";
            }
            var classAnonMatch = Regex.Match(content, @"\[AllowAnonymous\]\s*(?:\[[^\]]+\]\s*)*public\s+class\s+" + className);
            if (classAnonMatch.Success) classAuth = "[AllowAnonymous]";

            writer.WriteLine($"## {className}");
            if (!string.IsNullOrEmpty(classRoute)) writer.WriteLine($"**Base Route:** {classRoute}");
            writer.WriteLine($"**Class Level Authorization:** {classAuth}");
            writer.WriteLine();
            writer.WriteLine("| Method | HTTP Verb | Route | Authorization | Access Control Explanation |");
            writer.WriteLine("|---|---|---|---|---|");

            var methodRegex = new Regex(@"(?:\[(HttpGet|HttpPost|HttpPut|HttpDelete|HttpPatch)(?:\(""(.*?)""\))?\]\s*)+(?:\[(.*?)\]\s*)*public\s+(?:async\s+)?(?:Task<.*?>|IActionResult|ActionResult<.*?>|Task)\s+(\w+)\(");
            var matches = methodRegex.Matches(content);

            var methodInfos = new List<string>();

            // To get auth attributes, we can extract the block of attributes preceding the method
            // We can match everything from the previous '}' or ';' up to the method name.
            // Let's iterate lines to be safer.
            var lines = File.ReadAllLines(file);
            string currentHttpVerb = "";
            string currentRoute = "";
            string currentAuth = "";
            string currentMethodName = "";
            bool inMethodAttrs = false;

            for (int i = 0; i < lines.Length; i++)
            {
                var line = lines[i].Trim();
                if (line.StartsWith("[Http"))
                {
                    inMethodAttrs = true;
                    var verbMatch = Regex.Match(line, @"\[(HttpGet|HttpPost|HttpPut|HttpDelete|HttpPatch)");
                    if (verbMatch.Success) currentHttpVerb = verbMatch.Groups[1].Value.Replace("Http", "").ToUpper();
                    
                    var routeMatch = Regex.Match(line, @"\(""(.*?)""\)");
                    if (routeMatch.Success) currentRoute = routeMatch.Groups[1].Value;
                }
                else if (inMethodAttrs && line.StartsWith("[Authorize"))
                {
                    currentAuth = line;
                }
                else if (inMethodAttrs && line.StartsWith("[AllowAnonymous"))
                {
                    currentAuth = "[AllowAnonymous]";
                }
                else if (inMethodAttrs && line.StartsWith("public"))
                {
                    var nameMatch = Regex.Match(line, @"public\s+(?:async\s+)?(?:Task(?:<.*?>)?|IActionResult|ActionResult(?:<.*?>)?|\w+)\s+(\w+)\(");
                    if (nameMatch.Success)
                    {
                        currentMethodName = nameMatch.Groups[1].Value;
                        string finalAuth = string.IsNullOrEmpty(currentAuth) ? (string.IsNullOrEmpty(classAuth) ? "None" : classAuth) : currentAuth;
                        string fullRoute = classRoute + (string.IsNullOrEmpty(currentRoute) ? "" : "/" + currentRoute);
                        
                        string explanation = "";
                        if (finalAuth == "[AllowAnonymous]") explanation = "Public access, no token required.";
                        else if (finalAuth == "[Authorize]" || finalAuth == "None") explanation = "Requires valid authentication token.";
                        else if (finalAuth.Contains("Roles")) 
                        {
                            var rolesMatch = Regex.Match(finalAuth, @"Roles\s*=\s*""(.*?)""");
                            if (rolesMatch.Success)
                                explanation = $"Requires valid token AND user must have one of these roles: {rolesMatch.Groups[1].Value}";
                        }
                        
                        writer.WriteLine($"| {currentMethodName} | {currentHttpVerb} | {fullRoute.Replace("//", "/")} | {finalAuth} | {explanation} |");
                    }
                    // Reset
                    currentHttpVerb = ""; currentRoute = ""; currentAuth = ""; currentMethodName = ""; inMethodAttrs = false;
                }
                else if (line.StartsWith("public") && line.Contains("(") && !line.Contains("class") && !line.Contains("interface"))
                {
                    // Reset if we hit another public method without http attributes
                    currentHttpVerb = ""; currentRoute = ""; currentAuth = ""; currentMethodName = ""; inMethodAttrs = false;
                }
            }
        }
    }
}
